flow/directives.py

```diff
@@ -0,0 +1,328 @@
+        appropriate value for these operations running in parallel. Defaults to
+        the sum of the two.
+    finalize : callable, optional:
+        A function that takes the current value of the directive and the
+        directives object it is a child of and outputs the finalized value for
+        that directive. This is useful if some directives have multiple ways to
+        be set or are dependent in some way on other directives. The default is
+        to just return the current set value.
+    """
+    def __init__(self, name, *, validation=None, default=None,
+                 serial=max, parallel=operator.add, finalize=None):
+        self.name = name
+        self.default = default
+        self.serial = serial
+        self.parallel = parallel
+
+        def identity(v):
+            return v
+
+        def default_finalize(v, directives):
+            return v
+
+        self.validation = identity if validation is None else validation
+        self.finalize = default_finalize if finalize is None else finalize
+
+    def __call__(self, value):
+        if callable(value):
+            return value
+        else:
+            return self.validation(value)
+
+
+class Directives(MutableMapping):
+    """Class that maps environment and user specified objects for execution.
+
+    Resource request and execution style directives are the primary purpose for
+    this class. Examples include the number of processes expected to run or the
+    time to request to run.
+
+    Parameters
+    ----------
+    available_directives_list : Sequence[_DirectivesItem]
+        The sequence of all environment-specified directives. All other
+        directives are user-specified and not validated. There is no way to
+        include environment directives not specified at initialization without
+        using the object's private API.
+    """
+    def __init__(self, available_directives_list):
+        self._directive_definitions = dict()
+        self._defined_directives = dict()
+        self._user_directives = dict()
+
+        for directive in available_directives_list:
+            self._add_directive(directive)
+
+    def _add_directive(self, directive):
+        if not isinstance(directive, _DirectivesItem):
+            raise TypeError("Expected a _DirectivesItem object. Received {}".format(
+                type(directive)))
+        elif directive.name in self._directive_definitions:
+            raise ValueError("Cannot define {} twice.".format(directive.name))
+        else:
+            self._directive_definitions[directive.name] = directive
+            self._defined_directives[directive.name] = directive.default
+
+    def _set_defined_directive(self, key, value):
+        try:
+            self._defined_directives[key] = self._directive_definitions[key](value)
+        except Exception as err:
+            raise err.__class__('Error setting directive "{}": {}'.format(key, str(err)))
```

Comment:
I would prefer to use `raise from` with a custom DirectivesError error class. Follow this example: https://stackoverflow.com/a/29442282

Search in the code base for an example error class (ideally one with a docstring), but something like this should work:
```python
class DirectivesError(Exception):
    pass

# Then raise from:
raise DirectivesError('Error...') from err
```

Reply:
And if you _do_ want to forward the exception type, use `type(err)('Error...')` instead of `err.__class__('Error...')`.